Port to 1.21.7

Awakened-Redstone opened the pull request:
I'll keep it in draft for now as I want to make sure to test things first

TODO(fixes):
- [x]  Popup block title
- [x] item acceptor active is broken
- [x] Notes no longer seem to render in hand

Commit:
chore: Add shell.nix for nix users (me)

Commit:
chore: Update codecs and packet codecs

Commit:
chore: Add access widener

Commit:
chore: Remove temp veil compat

Commit:
chore: Disable EMI compat while it is not available on 1.21.7

Commit:
chore(port): Port block entities (NBT/data and others)

Commit:
chore(port): Port blocks

Commit:
chore: Use / on path append over .

Commit:
chore(port): Port items

Commit:
chore(port): Update gradle

Commit:
chore(port): Update versions

Commit:
chore: Bump mod version

Commit:
chore: Add Awakened Redstone to contributors

Commit:
port(client): Update item tint to use models models

Commit:
port(mixin): Update mixins

Commit:
port(rendering): Early rendering port

Commit:
port(others): datagen

Commit:
port(rendering): Baked BE rendering and other fixes

Commit:
port(block): Fix crash due to particles

Commit:
port(block): Fix block data

TheEpicBlock commented on src/main/java/dev/hephaestus/glowcase/block/GlowcaseBlock.java:
```diff
@@ -100,14 +105,28 @@ protected static <E extends BlockEntity, A extends BlockEntity> BlockEntityTicke
+				return player.isCreative() && player.canModifyAt(serverWorld, pos);
+			}
+
+			return player.isCreative();
```
Why is `canModifyAt` not being called for singleplayer worlds?

Awakened-Redstone replied:
it requires a server world now
a single-player world contains both worlds so it should still be called

TheEpicBlock commented on src/main/java/dev/hephaestus/glowcase/block/GlowcaseBlock.java:
```diff
@@ -100,14 +105,28 @@ protected static <E extends BlockEntity, A extends BlockEntity> BlockEntityTicke
+		return false;
+	}
+
+	public void appendTooltip(ItemStack stack, Item.TooltipContext context, TooltipDisplayComponent displayComponent, Consumer<Text> textConsumer, TooltipType type) {
```
This method is now one that's glowcase exclusive right? I think if we name it `appendGlowcaseTooltip` that'd be more clear (because I confused it for a mc one at first readthrough)

Awakened-Redstone replied:
i can do that
but i also plan to better check the new features to see why mojang removed it on blocks and marked the item one ad deprecated. i may move it into the lore component

TheEpicBlock commented on src/main/java/dev/hephaestus/glowcase/block/entity/ConfigLinkBlockEntity.java:
```diff
@@ -44,16 +46,18 @@ public void setUrl(String newUrl) {
+		super.readData(view);
+
+		this.title = view.getString("title", "");
+		this.url = view.getString("url", "glowcase:mod/modmenu");
```
Is this a new default? If so, that's great! Having a default helps explain the syntax (I would've never guessed it)

Awakened-Redstone replied:
yeah, the new data system gives an optional of t or t if you pass a default (unless it's a codec/read, that is always an optional with no way to pass a default, but you can use .orElse)

Ampflower replied:
If you're curious of why the weird syntax: it was intended to be extended later on to allow mods to register directly initializing a specific screen, and I didn't want to marry it to subpar weirdness of just `modmenu` being ambiguous.

Admittedly that should be a dedicated issue to discuss tho.

TheEpicBlock commented on src/main/java/dev/hephaestus/glowcase/client/ScreenImageCache.java:
```diff
@@ -250,11 +252,13 @@ public ScreenTexture(URL url) {
 					TextureManager textureManager = MinecraftClient.getInstance().getTextureManager();
-					this.texture = textureManager.registerDynamicTexture("glowcase/img", nativeTexture);
+					this.texture = Glowcase.id("glowcase/img", imageHash);
+					textureManager.registerTexture(this.texture, nativeTexture);
```
Does this change the behaviour? The first parameter used to be constant but is now different for each image

Awakened-Redstone replied:
I was unsure on how it worked, and imagined minecraft assigned you an id for each image, if it was all in a single texture I can change it back

I should give it a look when I get to fixing the other bugs

TheEpicBlock reviewed (approved):
Haven't been able to test it in-game yet but this all looks awesome. Seems like quite a lot had to be done for this port!

Tomate0613 committed:
Add back polygon offset and remove unused layering

Awakened-Redstone committed:
Add files via upload

Tomate0613 committed:
fix popup block title not rendering

Tomate0613 committed:
fix first person item rendering

Awakened-Redstone commented on src/main/java/dev/hephaestus/glowcase/client/render/block/entity/PopupBlockEntityRenderer.java:
```diff
@@ -37,7 +37,7 @@ public void render(PopupBlockEntity entity, float f, MatrixStack matrices, Verte
 			// Fixes shadow being rendered in front of actual text
 			matrices.scale(1, 1, -1);
-			context.getTextRenderer().draw(entity.title, 0, 0, 0xFFFFFF, true, matrices.peek().getPositionMatrix(), vertexConsumers, TextLayerType.NORMAL, 0, LightmapTextureManager.MAX_LIGHT_COORDINATE);
+			context.getTextRenderer().draw(entity.title, 0, 0, 0xFFFFFFFF, true, matrices.peek().getPositionMatrix(), vertexConsumers, TextLayerType.NORMAL, 0, LightmapTextureManager.MAX_LIGHT_COORDINATE);
```
ah, right, alpha, I forgot that was changed

Tomate0613 committed:
Fix suggestion list widget

Tomate0613 committed:
Add helper factory methods

Commit:
chore: Change some TODOs to FIXME

Commit:
port(rendering): Fix suggestions widget blur (this code is really cursed)

Commit:
port(rendering): Fix suggestions widget blur (this code is really cursed)

Commit:
Merge remote-tracking branch 'origin/1.21.7' into 1.21.7

Commit:
chore: Remove depth buffer

Tomate0613 committed:
Cleanup

Ampflower committed:
fix: Check queue to respect sound block's delay

isPlaying was evidently not sufficient for checking if the sound was also queued.

Ampflower commented:
I'll mention; the datagen setup does not seem to be automated in any capacity yet, a clean build will be missing model definitions.

Awakened-Redstone commented:
> I'll mention; the datagen setup does not seem to be automated in any capacity yet, a clean build will be missing model definitions.

oh yeah, it's not pushed to the repo, should I push it or should I set some sort of automation?

Commit:
chore: Push datagen data to the branch

Commit:
Merge remote-tracking branch 'origin/1.21.7' into 1.21.7

HamaIndustries commented:
You can push it, but if you know how to set up automation that would be a nice addition

HamaIndustries commented:
Reposting from the thread. Everything else is fine as far as I can tell.

sound block: randomly selects one sound for multi-sound events and sticks to it rather than varying
hyperlink block: title doesn't seem to appear over the block when hovered
provider block: doesn't display countdown when in TIMED mode. would be nice to be able to interact with empty hand in creative. crouch click could give a stack in always mode.
tablet screen: n/a for fests. doesn't crash.
icon button: which block uses this?
text input widgets: (noticed this on particle block mainly): should auto highlight when clicked once. Also tabbing around would be great
resource location widget: can't tab/enter autocomplete or press up down arrows
vec3 fields: work but have floating point issues; 0.05 becomes 0.074505...

Since these are all minor issues, I think we're ready to ship.

Commit:
chore: Fix non rendering text

Commit:
chore: Deprecate glowcase block tooltip

Commit:
chore: Add QoL to item providers

Commit:
chore: Keep particle and sound blocks visible while targeted for easier handling
